### PR TITLE
Making grammar unambiguous

### PR DESCRIPTION
**Pull request by sauliusg**
In this PR I offer a Filter language grammar that is fully specified in EBNF, including the space handling and token definitions, is unambiguous and lends itself (hopefully) to efficient implementation using current compiler construction methodologies. Pls. review :)

### CODE REVIEW

**Comment by rartino** [comment]
Nice work. But, you need to also update the regular expressions in Appendix 3 to only allow lower-case letters now.

**Comment by rartino** [comment]
Two more things:
* Shouldn't you also update section 5.1 (or is the plan to do that in a separate PR?)
* Shouldn't Space now be = ' ' | '\t' | '\n' | '\r' | '\f' | '\v' (and perhaps be called WhiteSpace or WS?)

File: optimade.md [hunk 1]
```diff
@@ -1729,21 +1699,63 @@ LengthComparison = LENGTH, Identifier, Operator, Value ;
 
 IdentifierZipAddon = ':', Identifier, {':', Identifier} ;
```
Comment (rartino) [inline optimade.md]: Why are these ':' not Colon?

File: optimade.md [hunk 2]
```diff
@@ -1,4 +1,4 @@
-# OPTiMaDe API specification v0.9.8-develop
+ OPTiMaDe API specification v0.9.8-develop
```
Comment (merkys) [inline optimade.md]: Is this intentional?

Comment (sauliusg) [reply optimade.md]: No... Probably an accidental change.

Comment (sauliusg) [reply optimade.md]: Reinserted the '#' character back.

File: optimade.md [hunk 3]
```diff
@@ -1775,19 +1783,46 @@ Digits =  Digit, { Digit } ;
...
+
+UnicodeHighChar = ? [^\x00-\x7F] ? ;
+ 
 (* END EBNF GRAMMAR Number *)
 (* END EBNF GRAMMAR Filter *)
 ```
-Note: in the parsing of filters according to this grammar, all whitespace (space, tabs, newlines) should be
-discarded between tokens.
+
+Note: whe implementing a parser according this grammar, the
```
Comment (rartino) [inline optimade.md]:
```suggestion
Note: when implementing a parser according this grammar, the
```

File: optimade.md [hunk 4]
```diff
@@ -1775,19 +1783,46 @@ Digits =  Digit, { Digit } ;
...
+implementers MAY choose to construct a lexer that ignores all
+whitespace (space, tabs, newlines, vertical tabulation an format feed
```
Comment (rartino) [inline optimade.md]:
```suggestion
whitespace (space, tabs, newlines, vertical tabulation and form feed
```

File: optimade.md [hunk 5]
```diff
@@ -1775,19 +1783,46 @@ Digits =  Digit, { Digit } ;
...
+characters, as described in the grammar 'Space' definition), and use
+such lexer to recognise language elements that are described in the
```
Comment (rartino) [inline optimade.md]:
```suggestion
such a lexer to recognize language elements that are described in the
```

File: optimade.md [hunk 6]
```diff
@@ -990,9 +990,10 @@ Or, fully URL encoded :
 The following tokens are used in the filter query component:
 
 * **Property names** (see section [6. Entry List](#h.6)): Are to follow the identifier
-  syntax of programming languages -- the first character MUST be a letter, the subsequent symbols
-  MUST be alphanumeric; the underscore ("\_", ASCII 95 dec (0x5F)) is considered to be a letter.
-  Identifiers are case-sensitive. The length of the identifiers is not limited, except that when passed as a
+  syntax of programming languages -- the first character MUST be a lowercase letter, the subsequent symbols
```
Comment (rartino) [inline optimade.md]: Isn't it a bit weird now to keep the "follow the identifier syntax of programming languages", when most programming language allows mixed-case identifiers?

Comment (sauliusg) [reply optimade.md]: I agree, it may be more confusing than clarifying. Probably we should remove reference to "most programming languages" at all.

Comment (sauliusg) [reply optimade.md]: Re-wrote the "most programming languages" paragraph.

**Comment by rartino** [comment]
> The PR seems OK to me. Just a single comment: is the change to the specs heading (line 1 of optimade.md) intentional?

@merkys This is fixed now, right?

**Pull request merged**